testkit/src/lib.rs

```diff
@@ -16,106 +16,86 @@
-//! use exonum_merkledb::{Snapshot, Fork};
-//! use exonum_testkit::{ApiKind, TestKitBuilder};
+//! ```
+//! use exonum::{
+//!     runtime::{InstanceDescriptor, rust::{Transaction, TransactionContext, Service}},
+//!     blockchain::{Block, Schema, ExecutionError, InstanceCollection},
+//!     crypto::{gen_keypair, Hash},
+//!     explorer::TransactionInfo,
+//!     helpers::Height,
+//!     api::node::public::explorer::{BlocksQuery, BlocksRange, TransactionQuery},
+//! };
+//! use serde_derive::{Serialize, Deserialize};
+//! use exonum_derive::{exonum_service, ServiceFactory, ProtobufConvert};
+//! use exonum_merkledb::{ObjectHash, Snapshot};
+//! use exonum_testkit::{txvec, ApiKind, TestKitBuilder};
 //!
 //! // Simple service implementation.
 //!
-//! const SERVICE_ID: u16 = 1;
+//! const SERVICE_ID: u32 = 1;
 //!
 //! #[derive(Debug, Clone, Serialize, Deserialize, ProtobufConvert)]
 //! #[exonum(pb = "exonum_testkit::proto::examples::TxTimestamp")]
-//! struct TxTimestamp {
+//! pub struct TxTimestamp {
 //!     message: String,
 //! }
 //!
-//! #[derive(Debug, Clone, Serialize, Deserialize, TransactionSet)]
-//! enum TimestampingTransactions {
-//!     TxTimestamp(TxTimestamp),
-//! }
-//!
-//! impl TxTimestamp {
-//!    fn sign(author: &PublicKey, msg: &str, key: &SecretKey) -> Verified<AnyTx> {
-//!        let tx = TxTimestamp{ message: msg.to_owned() };
-//!        Message::sign_transaction(tx, SERVICE_ID, *author, key)
-//!    }
-//! }
-//!
+//! #[derive(Clone, Default, Debug, ServiceFactory)]
+//! #[exonum(
+//!     artifact_name = "timestamping",
+//!     artifact_version = "1.0.0",
+//!     proto_sources = "exonum_testkit::proto",
+//!     implements("TimestampingInterface")
+//! )]
 //! struct TimestampingService;
 //!
-//! impl Transaction for TxTimestamp {
-//!     fn execute(&self, _context: TransactionContext) -> ExecutionResult {
-//!         Ok(())
-//!     }
-//! }
-//!
 //! impl Service for TimestampingService {
-//!     fn service_name(&self) -> &str {
-//!         "timestamping"
-//!     }
-//!
-//!     fn state_hash(&self, _: &Snapshot) -> Vec<Hash> {
-//!         Vec::new()
-//!     }
+//!     fn state_hash(&self, _: InstanceDescriptor, _: &dyn Snapshot) -> Vec<Hash> { vec![] }
+//! }
 //!
-//!     fn service_id(&self) -> u16 {
-//!         SERVICE_ID
-//!     }
+//! #[exonum_service]
+//! pub trait TimestampingInterface {
+//!     fn timestamp(&self, _: TransactionContext, arg: TxTimestamp) -> Result<(), ExecutionError>;
+//! }
 //!
-//!     fn tx_from_raw(&self, raw: AnyTx) -> Result<Box<Transaction>, failure::Error> {
-//!         let tx = TimestampingTransactions::tx_from_raw(raw)?;
-//!         Ok(tx.into())
+//! impl TimestampingInterface for TimestampingService {
+//!     fn timestamp(&self, _: TransactionContext, arg: TxTimestamp) -> Result<(), ExecutionError> {
+//!         Ok(())
 //!     }
 //! }
 //!
 //! fn main() {
 //!     // Create testkit for network with four validators.
```

Comment:
Maybe add a note like "...and add a builtin `timestamping` service with ID=1"?

Reply:
Done